core/src/introspection.cpp

```diff
@@ -64,34 +64,74 @@ std::string getTaskId(const TaskPrivate* task) {
 }
 }  // namespace
 
+/**
+ * A shared executor between all tasks' introspection, this class will keep track of the number of the added nodes and
+ * will only create a spinning thread when we have nodes in the executor, once all the nodes are removed from the
+ * executor it will stop the spinning thread, later on if a new node is added a spinning thread will be started again
+ */
+class IntrospectionExecutor
+{
+public:
+	void add_node(const rclcpp::Node::SharedPtr& node) {
+		std::call_once(once_flag_, [this] { executor_ = rclcpp::executors::SingleThreadedExecutor::make_unique(); });
+		std::lock_guard<std::mutex> lock(mutex_);
+		executor_->add_node(node);
+		if (nodes_count_ == 0)
+			executor_thread_ = std::thread([this] { executor_->spin(); });
+		++nodes_count_;
+	}
+
+	void remove_node(const rclcpp::Node::SharedPtr& node) {
+		std::lock_guard<std::mutex> lock(mutex_);
+		executor_->remove_node(node);
+		if (--nodes_count_ == 0) {
+			executor_->cancel();
+			if (executor_thread_.joinable())
+				executor_thread_.join();
+		}
+	}
+
+private:
+	rclcpp::executors::SingleThreadedExecutor::UniquePtr executor_;
+	std::thread executor_thread_;
+	size_t nodes_count_ = 0;
+	std::mutex mutex_;
+	std::once_flag once_flag_;
+};
+
 class IntrospectionPrivate
 {
 public:
 	IntrospectionPrivate(const TaskPrivate* task, Introspection* self)
-	  : nh_(std::string("~/") + task->ns())  // topics + services are advertised in private namespace
-	  , task_(task)
-	  , task_id_(getTaskId(task)) {
-		task_description_publisher_ =
-		    nh_.advertise<moveit_task_constructor_msgs::TaskDescription>(DESCRIPTION_TOPIC, 2, true);
+	  : task_(task)
+	  , task_id_(getTaskId(task))
+	  , node_(rclcpp::Node::make_shared(task_id_ + "_introspection", task_->ns())) {
+		executor_.add_node(node_);
+		task_description_publisher_ = node_->create_publisher<moveit_task_constructor_msgs::msg::TaskDescription>(
+		    DESCRIPTION_TOPIC, rclcpp::QoS(2).transient_local());
 		// send reset message as early as possible to give subscribers time to see it
 		indicateReset();
 
-		task_statistics_publisher_ =
-		    nh_.advertise<moveit_task_constructor_msgs::TaskStatistics>(STATISTICS_TOPIC, 1, true);
-		solution_publisher_ = nh_.advertise<moveit_task_constructor_msgs::Solution>(SOLUTION_TOPIC, 1, true);
-
-		get_solution_service_ =
-		    nh_.advertiseService(std::string(GET_SOLUTION_SERVICE "_") + task_id_, &Introspection::getSolution, self);
+		task_statistics_publisher_ = node_->create_publisher<moveit_task_constructor_msgs::msg::TaskStatistics>(
+		    STATISTICS_TOPIC, rclcpp::QoS(1).transient_local());
```

Comment:
This's how to do latching in ROS2 [see](https://index.ros.org/doc/ros2/Concepts/About-Quality-of-Service-Settings/#comparison-to-ros-1)

Reply:
Cool! Could you add a short note for this?